<launch>
<arg name="input_bag_name" default="box_2rs_bag2.bag" />
    <node pkg="rosbag" type="play" name="player" output="screen" args="/home/giacomo/inspectrone_ws/bags/box_rs/$(arg input_bag_name) --clock -s 22"/>
    
    <!-- Set up transformation 2 realsense t265 in the box-->
    <node name="odom1_tf_listener" output="screen" pkg="InspectroneRobotLocalization" type="publish_odom_rs_test_diff" required="true">
        <param name="camera" type="str" value="camera1" />
        <param name="base_link" type="str" value="base_link_1" />
        <param name="prefix_odom" value="cam1" />
        <param name="differential" type="bool" value="true" />
    </node>
        
    <node name="odom2_tf_listener" output="screen" pkg="InspectroneRobotLocalization" type="publish_odom_rs_test_diff" required="true">
        <param name="camera" type="str" value="camera2" />
        <param name="base_link" type="str" value="base_link_2" />
        <param name="prefix_odom" value="cam2" />
        <param name="differential" type="bool" value="true" />
    </node>

    <param name="use_sim_time" value="true" />
    <!-- static_transform_publisher x y z yaw pitch roll frame_id child_frame_id period_in_ms-->
    <node pkg="tf" type="static_transform_publisher" name="static_cam1_pose_to_bl_1" args="-0.0265 0.0 0.0  1.57 0.0 1.57 cam1_pose_frame base_link_11 1000" />
    <node pkg="tf" type="static_transform_publisher" name="static_odom_to_cam2_odom" args="0.0 0.0 0.0265 0.0 -1.57 -1.57 odom cam1_odom_frame 1000" />

    <node name="ekf_odom" pkg="robot_localization" type="ekf_localization_node" clear_params="true" required="true">
            <param name="sensor_timeout" value="1"/>
            <!--param name="map_frame" value="map" /-->
            <param name="odom_frame" value="odom" />
            <param name="base_link_frame" value="base_link" />
            <param name="world_frame" value="odom" />

            <param name="frequency" value="20" />

            <param name="odom0" value="/cam1/odom/sample_throttled" />
            <!-- <param name="odom1" value="/cam2/odom/sample_throttled" /> -->

            <!-- X,Y,Z,roll,pitch,yaw,X˙,Y˙,Z˙,roll˙,pitch˙,yaw˙,X¨,Y¨,Z¨)-->
            <rosparam param="odom0_config">[
                                            false, false, false,
                                            true, true, true,
                                            true, true, true,
                                            false, false, false,
                                            false, false, false]</rosparam>

            
            <!--<rosparam param="imu0_config">[false, false, false,
                                            false, false, false,
                                            false, false, false,
                                            true, true, true,
                                            true, true, true]</rosparam>
            <rosparam param="imu1_config">[false, false, false,
                                            false, false, false,
                                            false, false, false,
                                            true, true, true,
                                            true, true, true]</rosparam>-->

            <param name="odom0_differential" value="true" />
            
            
            <!-- <rosparam param="odom1_config">[
                                            true, true, true,
                                            true, true, true,
                                            true, true, true,
                                            false, false, false,
                                            false, false, false]</rosparam>
            <param name="odom1_differential" value="true" />
            <param name="odom1_relative" value="false" />
            <param name="odom1_queue_size" value="50" /> -->


            <!-- mahalanobis distance threshold for odom0 and odom1 -->
            <!--<param name="odom0_pose_rejection_threshold" value="5" />
            <param name="odom1_pose_rejection_threshold" value="5" />-->
            <!--<param name="odom0_twist_rejection_threshold" value="10" />
            <param name="odom1_twist_rejection_threshold" value="10" />-->

            <param name="odom0_relative" value="false" />

            <param name="odom0_queue_size" value="50" />

            <param name="print_diagnostics" value="true"/>

            <param name="debug"           value="true"/>
            <param name="debug_out_file"  value="debug_ekf_localization.txt"/>
            <!-- If set to true will not run the filter on start. -->
            <!--<param name="disabled_at_startup" value="true" />-->
            
            <!-- If true, will dynamically scale the process_noise_covariance based on the robot’s velocity. This is useful, e.g., when you want your robot’s estimate error covariance to stop growing when the robot is stationary.-->
            <param name="dynamic_process_noise_covariance" value="true" />

            <!-- The process noise covariance, commonly denoted Q, is used to model uncertainty in the prediction stage of the filtering algorithms. In general, the larger the value for Q relative to the variance for a given variable in an input message, the faster the filter will converge to the value in the measurement.-->
            <!-- EKF param - trials -->
            <rosparam param="process_noise_covariance">[1e-2, 0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                        0,    1e-2, 0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                        0,    0,    1e-2, 0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                        0,    0,    0,    1e-2 ,0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                        0,    0,    0,    0,    1e-2, 0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                        0,    0,    0,    0,    0,    1e-2, 0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    1e-1,  0,     0,    0,    0,    0,    0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,     1e-1,  0,    0,    0,    0,    0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,     0,     1e-1, 0,    0,    0,    0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,     0,     0,    1e-5, 0,    0,    0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    1e-5, 0,    0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    1e-5, 0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    1e-5, 0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    1e-5, 0,
                                                        0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    1e-5]</rosparam>
        <!-- initial P -->
        <!--<rosparam param="initial_estimate_covariance">[1e-12, 0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                        0,    1e-12, 0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                        0,    0,    1e-12, 0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                        0,    0,    0,    1e-9, 0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                        0,    0,    0,    0,    1e-9, 0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                        0,    0,    0,    0,    0,    1e-9, 0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    1e-9, 0,    0,    0,     0,     0,     0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,    1e-9, 0,    0,     0,     0,     0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,    0,    1e-9, 0,     0,     0,     0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,    0,    0,    1e-9,  0,     0,     0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     1e-9,  0,     0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     1e-9,  0,    0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     1e-12, 0,    0,
                                                        0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    1e-12, 0,
                                                        0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    1e-12]</rosparam>-->
        <rosparam param="initial_state">[0.0,  0.0,  0.0,
                                        0.0,  0.0,  0.0,
                                        0.0,  0.0,  0.0,
                                        0.0,  0.0,  0.0,
                                        0.0,  0.0,  0.0]</rosparam>





    </node>
    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find InspectroneRobotLocalization)/rviz/box_odom_visualization.rviz"/>
</launch>